ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/ForkChoiceUtil.java

```diff
@@ -318,11 +319,13 @@ public static BlockImportResult on_block(
       try {
         if (justifiedCheckpoint.getEpoch().compareTo(store.getBestJustifiedCheckpoint().getEpoch())
             > 0) {
+          // TODO: Fork Choice Spec does not necessarily ask us to store this checkpoint state
           storeCheckpointState(
               store, st, justifiedCheckpoint, store.getBlockState(justifiedCheckpoint.getRoot()));
           store.setBestJustifiedCheckpoint(justifiedCheckpoint);
         }
         if (should_update_justified_checkpoint(store, justifiedCheckpoint)) {
+          // TODO: Fork Choice Spec does not necessarily ask us to store this checkpoint state
```

Comment:
We expect that any justified checkpoint has a checkpoint state stored for it. Otherwise we'll throw `NullPointerException` when later processing attestations in `get_latest_attesting_balance`:
`BeaconState state = store.getCheckpointState(store.getJustifiedCheckpoint());`

The fork choice spec actually does this store because it says to run `on_attestation` on every received attestation whether by gossip included in a block.  Apart from the validation checks for attestations, that ensures that the target state is in `store.checkpoint_states` and that `store.latest_messages` gets updates based on the attestation.

So storing these checkpoint states is part of the `on_attestation` processing.  The other half is updating latest messages which a ticket is open for (but unclear on how important it is to do when we're syncing such old blocks).

Technically, we're storing fewer checkpoint states that the spec implementation.  `on_attestation` would store the checkpoint state for the target of any attestation, we only store it when there are enough attestations to make it the best justified or justified state.

Reply:
Makes sense.